File: Scripts/ErrorHandler.vb
Option Strict On
Option Explicit On

Imports System.Windows.Forms
Imports Excel = Microsoft.Office.Interop.Excel

Namespace Scripts

    Public Class ErrorHandler

        Public Shared Sub DisplayMessage(ex As Exception)
            Dim sf As New System.Diagnostics.StackFrame(1)
            Dim caller As System.Reflection.MethodBase = sf.GetMethod()
            Dim currentProcedure As String = (caller.Name).Trim()
            Dim errorMessageDescription As String = ex.ToString()
            errorMessageDescription = System.Text.RegularExpressions.Regex.Replace(errorMessageDescription, "\r\n+", " ")
            Dim msg As String = "Contact your system administrator. A record has been created in the log file." + Environment.NewLine
            msg += (Convert.ToString("Procedure: ") & currentProcedure) + Environment.NewLine
            msg += "Description: " + ex.ToString() + Environment.NewLine
            MessageBox.Show(msg, "Unexpected Error", MessageBoxButtons.OK, MessageBoxIcon.[Error])
        End Sub

        Public Shared Function IsActiveDocument(Optional showMsg As Boolean = False) As Boolean
            Try
                If Globals.ThisAddIn.Application.ActiveWorkbook Is Nothing Then
                    If showMsg = True Then
                        MessageBox.Show("The command could not be completed.  Please open a document and select a range.", My.Application.Info.Description, MessageBoxButtons.OK, MessageBoxIcon.Exclamation)
                    End If
                    Return False
                Else
                    Return True
                End If

            Catch ex As Exception
                ErrorHandler.DisplayMessage(ex)
                Return False
            End Try

        End Function

        Public Shared Function IsActiveSelection(Optional showMsg As Boolean = False) As Boolean
            Dim checkRange As Excel.Range = Nothing
            Try
                checkRange = TryCast(Globals.ThisAddIn.Application.Selection, Excel.Range)
                'must cast the selection as range or errors
                If checkRange Is Nothing Then
                    If showMsg = True Then
                        MessageBox.Show("The command could not be completed by using the range specified.  Select a single cell within the range and try the command again. [Range]", My.Application.Info.Description, MessageBoxButtons.OK, MessageBoxIcon.Exclamation)
                    End If
                    Return False
                Else
                    Return True
                End If

            Catch ex As Exception
                ErrorHandler.DisplayMessage(ex)
                Return False

            Finally
                If checkRange IsNot Nothing Then
                    'Marshal.ReleaseComObject(checkRange)
                End If
            End Try

        End Function

        Private Shared Function IsInCellEditingMode(Optional showMsg As Boolean = False) As Boolean
            Dim flag As Boolean = False
            Try
                'This will throw an Exception if Excel is in Cell Editing Mode
                Globals.ThisAddIn.Application.DisplayAlerts = False

            Catch generatedExceptionName As Exception
                If showMsg = True Then
                    MessageBox.Show("The procedure can not run while you are editing a cell.", "No action taken.", MessageBoxButtons.OK, MessageBoxIcon.Information)
                End If
                flag = True
            End Try
            Return flag

        End Function

        Public Shared Function IsEnabled(Optional showMsg As Boolean = False) As Boolean
            Try
                If IsActiveDocument(showMsg) = False Then
                    Return False
                Else
                    If IsActiveSelection(showMsg) = False Then
                        Return False
                    Else
                        If IsInCellEditingMode(showMsg) = True Then
                            Return False
                        Else
                            Return True
                        End If
                    End If
                End If

            Catch ex As Exception
                ErrorHandler.DisplayMessage(ex)
                Return False
            End Try

        End Function

    End Class

End Namespace

File: CS/Scripts/ErrorHandler.cs
using System;
using System.Windows.Forms;
using System.Runtime.InteropServices;
using Excel = Microsoft.Office.Interop.Excel;
using log4net;
using log4net.Config;

[assembly: log4net.Config.XmlConfigurator(Watch = true)]

namespace Favorites.Scripts
{
    public class ErrorHandler
    {
        private static readonly ILog log = LogManager.GetLogger(typeof(ErrorHandler));

        public static void SetLogPath()
        {
            XmlConfigurator.Configure();
            log4net.Repository.Hierarchy.Hierarchy h = (log4net.Repository.Hierarchy.Hierarchy)LogManager.GetRepository();
            string logFileName = System.IO.Path.Combine(Properties.Settings.Default.App_PathLocalData, AssemblyInfo.Product + ".log");
            foreach (var a in h.Root.Appenders)
            {
                if (a is log4net.Appender.FileAppender)
                {
                    if (a.Name.Equals("FileAppender"))
                    {
                        log4net.Appender.FileAppender fa = (log4net.Appender.FileAppender)a;
                        fa.File = logFileName;
                        fa.ActivateOptions();
                    }
                }
            }
        }

        public static void CreateLogRecord()
        {
            try
            {
                System.Diagnostics.StackFrame sf = new System.Diagnostics.StackFrame(1);
                System.Reflection.MethodBase caller = sf.GetMethod();
                string currentProcedure = (caller.Name).Trim();
                log.Info("[PROCEDURE]=|" + currentProcedure + "|[USER NAME]=|" + Environment.UserName + "|[MACHINE NAME]=|" + Environment.MachineName);
            }
            catch (Exception ex)
            {
                ErrorHandler.DisplayMessage(ex);

            }
        }

        public static void DisplayMessage(Exception ex, Boolean isSilent = false)
        {
            System.Diagnostics.StackFrame sf = new System.Diagnostics.StackFrame(1);
            System.Reflection.MethodBase caller = sf.GetMethod();
            string currentProcedure = (caller.Name).Trim();
            string currentFileName = AssemblyInfo.GetCurrentFileName();
            string errorMessageDescription = ex.ToString();
            errorMessageDescription = System.Text.RegularExpressions.Regex.Replace(errorMessageDescription, @"\r\n+", " "); //the carriage returns were messing up my log file
            string msg = "Contact your system administrator. A record has been created in the log file." + Environment.NewLine;
            msg += "Procedure: " + currentProcedure + Environment.NewLine;
            msg += "Description: " + ex.ToString() + Environment.NewLine;
            log.Error("[PROCEDURE]=|" + currentProcedure + "|[USER NAME]=|" + Environment.UserName + "|[MACHINE NAME]=|" + Environment.MachineName + "|[FILE NAME]=|" + currentFileName + "|[DESCRIPTION]=|" + errorMessageDescription);
            if (isSilent == false)
            {
                MessageBox.Show(msg, "Unexpected Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
            }
        }

        public static bool IsActiveDocument(bool showMsg = false)
        {
            try
            {
                if (Globals.ThisAddIn.Application.ActiveWorkbook == null)
                {
                    if (showMsg == true)
                    {
                        MessageBox.Show("The command could not be completed.  Please open a document and select a range.", AssemblyInfo.Description, MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
                    }
                    return false;
                }
                else
                {
                    return true;
                }
            }
            catch (Exception ex)
            {
                ErrorHandler.DisplayMessage(ex);
                return false;
            }
        }

        public static bool IsActiveSelection(bool showMsg = false)
        {
            Excel.Range checkRange = null;
            try
            {
                checkRange = Globals.ThisAddIn.Application.Selection as Excel.Range; //must cast the selection as range or errors
                if (null == checkRange)
                {
                    if (showMsg == true)
                    {
                        MessageBox.Show("The command could not be completed by using the range specified.  Select a single cell within the range and try the command again. [Range]", AssemblyInfo.Description, MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
                    }
                    return false;
                }
                else
                {
                    return true;
                }
            }
            catch (Exception ex)
            {
                ErrorHandler.DisplayMessage(ex);
                return false;
            }
            finally
            {
                if (checkRange != null)
                    Marshal.ReleaseComObject(checkRange);
            }
        }

        public static bool IsValidListObject(bool showMsg = false)
        {
            Excel.ListObject tbl = null;
            try
            {
                tbl = Globals.ThisAddIn.Application.ActiveCell.ListObject;  // directly after the table is created this is not true
                if ((tbl == null))
                {
                    if (showMsg == true)
                    {
                        MessageBox.Show("The command could not be completed by using the range specified.  Select a single cell within the range and try the command again. [ListObject]", AssemblyInfo.Description, MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
                    }
                    return false;
                }
                else
                {
                    return true;
                }
            }
            catch (Exception)
            {
                return false;
            }
            finally
            {
                if (tbl != null)
                    Marshal.ReleaseComObject(tbl);
            }
        }

        private static bool IsInCellEditingMode(bool showMsg = false)
        {
            bool flag = false;
            try
            {
                Globals.ThisAddIn.Application.DisplayAlerts = false; //This will throw an Exception if Excel is in Cell Editing Mode
            }
            catch (Exception)
            {
                if (showMsg == true)
                {
                    MessageBox.Show("The procedure can not run while you are editing a cell.", "No action taken.", MessageBoxButtons.OK, MessageBoxIcon.Information);
                }
                flag = true;
            }
            return flag;
        }

        public static bool IsEnabled(bool showMsg = false)
        {
            try
            {
                if (IsActiveDocument(showMsg) == false)
                {
                    return false;
                }
                else
                {
                    if (IsActiveSelection(showMsg) == false)
                    {
                        return false;
                    }
                    else
                    {
                        if (IsInCellEditingMode(showMsg) == true)
                        {
                            return false;
                        }
                        else
                        {
                            return true;
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                ErrorHandler.DisplayMessage(ex);
                return false;
            }
        }

        public static bool IsAvailable(bool showMsg = false)
        {
            try
            {
                if (IsEnabled(showMsg) == false)
                {
                    return false;
                }
                else
                {
                    if (IsValidListObject(showMsg) == false)
                    {
                        return false;
                    }
                    else
                    {
                        return true;
                    }
                }
            }
            catch (Exception ex)
            {
                ErrorHandler.DisplayMessage(ex);
                return false;
            }
        }

        public static bool IsDate(object expression)
        {
            if (expression != null)
            {
                if (expression is DateTime)
                {
                    return true;
                }
                if (expression is string)
                {
                    DateTime time1;
                    return DateTime.TryParse((string)expression, out time1);
                }
            }
            return false;
        }

    }
}